kategory-docs/docs/docs/datatypes/statet/README.md

```diff
@@ -6,4 +6,142 @@ permalink: /docs/datatypes/statet/
+
+In the most basic of scenarios, we'll only be dealing with one monad at a time making our lives nice and easy. However, it's not uncommon to get into scenarios where some function calls will return `Either<Error, A>`, and others will return `State<S, A>`.
+
+So let's rewrite the example of [`State` docs]({{ '/docs/datatypes/state' | relative_url }}), but instead of representing the `Stack` as an optional `NonEmptyList` let's represent it as a `List`.
+
+```kotlin:ank:silent
+import kategory.*
+
+typealias Stack = List<String>
+
+fun pop(stack: Stack): Tuple2<Stack, String> = stack.first().let {
+    stack.drop(1) toT it
+}
+
+fun push(s: String, stack: Stack): Tuple2<Stack, Unit> =
+        listOf(s, *stack.toTypedArray()) toT Unit
+
+fun stackOperations(stack: Stack): Tuple2<Stack, String> {
+    val (s1, _) = push("a", stack)
+    val (s2, _) = pop(s1)
+    return pop(s2)
+}
+```
+
+```kotlin:ank
+stackOperations(listOf("hello", "world", "!"))
+```
+
+But if we now `pop` an empty `Stack` it will result in `java.util.NoSuchElementException: List is empty.`.
+
+```kotlin
+_stackOperations(listOf()) //java.util.NoSuchElementException: List is empty.
+```
+
+Luckily Kategory offers some nice solutions [`Functional Error Handling` docs]({{ '/docs/patterns/error_handling' | relative_url }}).
+Now we can model our error domain with ease.
+
+```kotlin:ank:silent
+sealed class StackError {
+    object StackEmpty : StackError()
+}
+
+typealias StackEmpty = StackError.StackEmpty
+
+fun popE(stack: Stack): Either<StackError, Tuple2<Stack, String>> =
+        if (stack.isEmpty()) StackEmpty.left()
+        else stack.first().let {
+            stack.drop(1) toT it
+        }.right()
+
+fun pushE(s: String, stack: Stack): Either<StackError, Tuple2<Stack, Unit>> =
+        (listOf(s, *stack.toTypedArray()) toT Unit).right()
+
+fun stackOperationsE(stack: Stack): Either<StackError, Tuple2<Stack, String>> {
+    return pushE("a", stack).flatMap { (s1, _) ->
+        popE(s1).flatMap { (s2, _) ->
+            popE(s2)
+        }
+    }
+}
+```
+```kotlin:ank
+stackOperationsE(listOf("hello", "world", "!"))
+```
+```kotlin:ank
+stackOperationsE(listOf())
+```
+
+As is immediately clear, this code while properly modelling the errors, has become more complex but our signature now represents a simple `Stack` as a `List` with an error domain.
+Let's refactor our manual state management in the form of `(S) -> Tuple2<S, A>` to `State`.
+
+So what we want is a return type that represents `Either` a `StackError` or a certain `State` of `Stack.` When working with `State` we don't pass around `Stack` anymore, so there is no parameter to check if the `Stack` is empty.
+
+```kotlin:ank:silent
+fun _popS(): Either<StackError, StateT<IdHK, Stack, String>> = TODO()
+```
+
+The only thing we can do is handle this with `StateT`. We want to wrap `State` with `Either`. Left to right wrap `State` `T` (with) `Either` `StateT<EitherHK`.
```

Comment:
Not sure if that "Left to right wrap... sentence is easy to understand". I had to read it a couple of times :D

Reply:
People needs to get used to the generics syntax of transformers stack in Kotlin. There is nothing to fear let us write:
```kotlin
StateT<EitherKindPartial<L>, Stack, String>
```
This means a state transformation considering a value may be `L` or a `String` where the internal state is tracked in a Stack. Several effects encoded in that program. Useful programs will most of the time become concrete over transformer stacks like this one if type classes and HKTs land in the land and we can start writing in tagless style